crates/wasmtime/src/v128.rs

```diff
@@ -0,0 +1,145 @@
+use crate::store::StoreOpaque;
+use crate::{ValRaw, ValType, WasmTy};
+use std::cmp::Ordering;
+use std::fmt;
+
+/// Representation of a 128-bit vector type, `v128`, for WebAssembly.
+///
+/// This type corresponds to the `v128` type in WebAssembly and can be used with
+/// the [`TypedFunc`] API for example. This is additionally
+/// the payload of [`Val::V128`](crate::Val).
+///
+/// # Platform specifics
+///
+/// This type can currently only be used on x86_64 and AArch64 with the
+/// [`TypedFunc`] API. Rust does not have stable support on other platforms for
+/// this type so invoking functions with `v128` parameters requires the
+/// [`Func::call`](crate::Func::call) API (or perhaps
+/// [`Func::call_unchecked`](crate::Func::call_unchecked).
+///
+/// [`TypedFunc`]: crate::TypedFunc
+#[derive(Copy, Clone)]
+#[repr(transparent)]
+pub struct V128(Abi);
+
+// NB: this is why this type is only suitable with `TypedFunc` on some platforms
+// and no other. See the documentation for each platform for why each ABI is
+// chosen.
+cfg_if::cfg_if! {
+    if #[cfg(target_arch = "x86_64")] {
+        // x86 vectors are represented with XMM registers which are represented
+        // with the `__m128i` type. This type is considered a vector type for
+        // ABI purposes which is implemented by Cranelift.
+        type Abi = std::arch::x86_64::__m128i;
+    } else if #[cfg(target_arch = "aarch64")] {
+        // AArch64 uses vector registered which here is used with a vector type.
+        // Note that the specific type shouldn't matter too much but the choice
+        // of using a vector is the significant part.
+        type Abi = std::arch::aarch64::uint8x16_t;
+    } else if #[cfg(target_arch = "riscv64")] {
+        // RISC-V currently always passes all vector arguments indirectly in the
+        // ABI. Currently Rust has no stable means of representing this meaning
+        // that a 128-bit representation is chosen here but it can't be passed
+        // to WebAssembly for example.
```

Comment:
```suggestion
        // that a 128-bit representation is chosen here but it can't be passed directly
        // to WebAssembly, for example, and must instead be passed through an array-call trampoline.
```

right?